nrf-hal-common/src/pwm.rs

```diff
@@ -0,0 +1,1017 @@
+            }
+        }
+        self
+    }
+
+    /// Cofigures how a sequence is read from RAM and is spread to the compare register.
+    #[inline(always)]
+    pub fn set_load_mode(&self, mode: LoadMode) -> &Self {
+        self.pwm.decoder.modify(|_r, w| w.load().bits(mode.into()));
+        if mode == LoadMode::Waveform {
+            self.disable_channel(Channel::C3);
+        } else {
+            self.enable_channel(Channel::C3);
+        }
+        self
+    }
+
+    /// Returns how a sequence is read from RAM and is spread to the compare register.
+    #[inline(always)]
+    pub fn get_load_mode(&self) -> LoadMode {
+        match self.pwm.decoder.read().load().bits() {
+            0 => LoadMode::Common,
+            1 => LoadMode::Grouped,
+            2 => LoadMode::Individual,
+            3 => LoadMode::Waveform,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Selects operating mode of the wave counter.
+    #[inline(always)]
+    pub fn set_counter_mode(&self, mode: CounterMode) -> &Self {
+        self.pwm.mode.write(|w| w.updown().bit(mode.into()));
+        self
+    }
+
+    /// Returns selected operating mode of the wave counter.    
+    #[inline(always)]
+    pub fn get_counter_mode(&self) -> CounterMode {
+        match self.pwm.mode.read().updown().bit() {
+            false => CounterMode::Up,
+            true => CounterMode::UpAndDown,
+        }
+    }
+
+    /// Selects source for advancing the active sequence.
+    #[inline(always)]
+    pub fn set_step_mode(&self, mode: StepMode) -> &Self {
+        self.pwm.decoder.modify(|_r, w| w.mode().bit(mode.into()));
+        self
+    }
+
+    /// Returns selected source for advancing the active sequence.
+    #[inline(always)]
+    pub fn get_step_mode(&self) -> StepMode {
+        match self.pwm.decoder.read().mode().bit() {
+            false => StepMode::Auto,
+            true => StepMode::NextStep,
+        }
+    }
+
+    // Internal helper function that returns 15 bit duty cycle value.
+    #[inline(always)]
+    fn get_duty_on_value(&self, index: usize) -> u16 {
+        let val = self.duty[index];
+        let is_inverted = (val >> 15) & 1 == 0;
+        match is_inverted {
+            false => val,
+            true => self.get_max_duty() - (val & 0x7FFF),
+        }
+    }
+
+    // Internal helper function that returns 15 bit inverted duty cycle value.
+    #[inline(always)]
+    fn get_duty_off_value(&self, index: usize) -> u16 {
+        let val = self.duty[index];
+        let is_inverted = (val >> 15) & 1 == 0;
+        match is_inverted {
+            false => self.get_max_duty() - val,
+            true => val & 0x7FFF,
+        }
+    }
+
+    /// Sets duty cycle (15 bit) for all PWM channels.
+    pub fn set_duty_on_common(&self, duty: u16) {
+        compiler_fence(Ordering::SeqCst);
+        unsafe {
+            *(self.duty.as_ptr() as *mut u16) = duty.min(self.get_max_duty()) & 0x7FFF;
```

Comment:
This is UB since `as_ptr` only gives immutable access.

```suggestion
            *(self.duty.as_mut() as *mut u16) = duty.min(self.get_max_duty()) & 0x7FFF;
```

You also need to take a `&mut self` here.

(this pattern also comes up a lot in the functions below)

Reply:
The problem is i can't use a &mut self because i want to be able to split out those channel and group "proxies"...
Would you suggest i use a RefCell instead?

Reply:
Either that or somehow use a buffer per channel/group if that's possible.

Reply:
Unfortunately that's not possible because of the way the DMA is implemented...
So I'll go with a RefCell then :)